### ISSUE TITLE
Legacy API mode leaks subprocesses

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#650]
In GitLab by @csilvers on Mar 2, 2018, 16:41

*Please describe how you installed Flake8*

pip install git+git://github.com/pycqa/flake8.git@dd1e9d1cb7e9a232946c06aca1564d48d4d6f65e

*Please provide the exact, unmodified output of `flake8 --bug-report`*

```
{
  "dependencies": [
    {
      "dependency": "setuptools", 
      "version": "38.5.1"
    }
  ], 
  "platform": {
    "python_implementation": "CPython", 
    "python_version": "2.7.12", 
    "system": "Linux"
  }, 
  "plugins": [
    {
      "is_local": false, 
      "plugin": "LintFlake8Plugins", 
      "version": "1.0"
    }, 
    {
      "is_local": false, 
      "plugin": "mccabe", 
      "version": "0.6.1"
    }, 
    {
      "is_local": false, 
      "plugin": "per-file-ignores", 
      "version": "0.4"
    }, 
    {
      "is_local": false, 
      "plugin": "pycodestyle", 
      "version": "2.3.1"
    }, 
    {
      "is_local": false, 
      "plugin": "pyflakes", 
      "version": "1.5.0"
    }, 
    {
      "is_local": false, 
      "plugin": "smart-longline", 
      "version": "1.0"
    }
  ], 
  "version": "3.5.0"
}
```

*Please describe the problem or feature*

If you use the legacy API in certain modes, it will leak subprocesses.  This is because init_report() does not properly destruct `self._application.file_checker._manager`.

Here is my code:
```
        style_guide = flake8.api.legacy.get_style_guide(jobs='8')
        style_guide.init_report(MyReporter)
        style_guide.check_files(myfiles)
```
When I run this, I see 17 python processes running.  One is my app, and 8 are the subprocesses I'd expect for the jobs, but the other 8 are orphaned.

This is because checker.Manager() creates a pool of subprocesses at `__init__` time, but destroys them at the end of `run()`.  Thus there's an implicit contract that if you create a `Manager` you call `run` on it.

However, this is not the case when you call `init_report`, which has the following code:
```
        self._application.file_checker_manager = None
        self._application.make_file_checker_manager()
```
It destroys the old `Manager` without doing anything that would delete its pool.

The simplest fix is to change init_report to this:
```
        self._application.file_checker_manager._force_cleanup()
        self._application.file_checker_manager = None
        self._application.make_file_checker_manager()
```
This violates the API by calling a private function.  It also only papers over the problem.  Probably, `Manager` should have something like a `__del__` that calls `_force_cleanup()`.  You still need to manually delete the Manager when you're done with it (or depend on CPython to do it for you).

Alternately, you could create the pool in `run_parallel()` rather than `__init__()`.  It seems to me there's no downside to doing this, but maybe there's a good reason it's done in `__init__()`.

**Comment by asottile** [comment]
In GitLab by @csilvers on Mar 2, 2018, 20:53

mentioned in merge request !228

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Apr 15, 2018, 12:31

mentioned in commit d0db8497b97cdcb0207da1a3b994e91ba554896c

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Apr 15, 2018, 12:31

closed via commit a5573fc8643de1df2bfe7429c81fcb0de42bd784

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Apr 15, 2018, 12:31

closed via merge request !228
